### PR TITLE
feat: platform dependent front matter values

### PR DESCRIPTION
**Pull request by cerkiewny**
## Motivation

Solves #5805 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Manual tests if behavior changes,
changes to existing tests to check if config with platform values is changing correctly

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/facebook/docusaurus, and link to your PR here.)

### CODE REVIEW

**Comment by cerkiewny** [comment]
For now it only works with hide_title, just want to see if it is good direction for PR.
I will update it with all values used in theme later on today.

File: packages/docusaurus-plugin-content-docs/src/docFrontMatter.ts [hunk 1]
```diff
@@ -14,14 +14,29 @@ import {
...
+
 // NOTE: we don't add any default value on purpose here
 // We don't want default values to magically appear in doc metadatas and props
 // While the user did not provide those values explicitly
 // We use default values in code instead
 const DocFrontMatterSchema = Joi.object<DocFrontMatter>({
   id: Joi.string(),
   title: Joi.string().allow(''), // see https://github.com/facebook/docusaurus/issues/4591#issuecomment-822372398
-  hide_title: Joi.boolean(),
+  hide_title: platformDependant(Joi.boolean()),
```
Comment (Josh-Cena) [inline packages/docusaurus-plugin-content-docs/src/docFrontMatter.ts:39]: `platformDependant` => `platformDependent`

This applies to every place

File: packages/docusaurus-plugin-content-docs/src/docFrontMatter.ts [hunk 2]
```diff
@@ -14,14 +14,29 @@ import {
 } from '@docusaurus/utils-validation';
 import {DocFrontMatter} from './types';
 
+// NOTE: This creates a validator from given joi type,
+// that allows the object with any fields, as long as they are
+// of the given type, or just a raw type given
+
+const platformDependant = (joiType: any): any =>
```
Comment (Josh-Cena) [inline packages/docusaurus-plugin-content-docs/src/docFrontMatter.ts:21]: Type should be `Joi.AnySchema` although I wonder if it can't be narrowed to just boolean/number/string

File: packages/docusaurus-plugin-content-docs/src/docFrontMatter.ts [hunk 3]
```diff
@@ -14,14 +14,29 @@ import {
...
+  Joi.alternatives(
+    joiType,
+    Joi.object()
+      .keys({
+        mobile: joiType,
+        desktop: joiType,
+      })
+      .pattern(/./, joiType),
```
Comment (Josh-Cena) [inline packages/docusaurus-plugin-content-docs/src/docFrontMatter.ts:29]: Just `Joi.object().pattern(/./, joiType)` is fine

File: packages/docusaurus-utils/src/platformDependant.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+export type PlatformDependant<Type> = Type | Record<string, Type>;
```
Comment (Josh-Cena) [inline packages/docusaurus-utils/src/platformDependant.ts:8]: I don't see a need to create a file for this... `T | Record<string, T>` is not a very tedious type. Also, move this to `theme-common` because utils is for Node code, not client code.

File: packages/docusaurus-theme-classic/src/theme/hooks/usePlatformValue.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import type {PlatformDependant} from '@docusaurus/utils';
+
+function usePlatformValue<Type>(
+  option: PlatformDependant<Type> | undefined,
+): Type | undefined {
+  const windowSize = useWindowSize();
+  if (option === undefined) {
+    return undefined;
+  }
+  if ((option as Record<string, Type>)[`${windowSize}`] !== undefined) {
```
Comment (Josh-Cena) [inline packages/docusaurus-theme-classic/src/theme/hooks/usePlatformValue.ts:18]: 1. Use a type guard to handle the `Type` case before `Record<string, Type>`. and put the type guard in `theme-common/platformDependent`
2. No need to use template strings; `useWindowSize` is already returning a string
3. Need to handle `ssr` as a window size here

File: packages/docusaurus-theme-classic/src/theme/DocItem/index.tsx [hunk 6]
```diff
@@ -31,12 +32,13 @@ export default function DocItem(props: Props): JSX.Element {
     toc_max_heading_level: tocMaxHeadingLevel,
   } = frontMatter;
   const {description, title} = metadata;
+  const platformHideTitle = usePlatformValue<boolean>(hideTitle);
```
Comment (Josh-Cena) [inline packages/docusaurus-theme-classic/src/theme/DocItem/index.tsx:35]: The generic is not necessary. `<boolean>` should be inferred

**Review by Josh-Cena — CHANGES_REQUESTED** [review]
Thanks, looks quite promising, but the TypeScript part is not good enough. I can fix that later if you need help

File: packages/docusaurus-plugin-content-docs/src/docFrontMatter.ts (hunk 2 above)
Comment (cerkiewny) [reply packages/docusaurus-plugin-content-docs/src/docFrontMatter.ts:21]: In case we would like to have something like screen size dependent logos etc, the joiType can be more complex object type, it would be also good to probably run validation of it in .desktop etc options in config

**Comment by cerkiewny** [comment]
> 
> 
> Thanks, looks quite promising, but the TypeScript part is not good enough. I can fix that later if you need help

I am not typescript expert, come more from c++ background so still getting used to the TypeScript, if you could provide some tips what should I change I would use this opportunity to improve my skill :)

File: packages/docusaurus-plugin-content-docs/src/docFrontMatter.ts (hunk 2 above)
Comment (Josh-Cena) [reply packages/docusaurus-plugin-content-docs/src/docFrontMatter.ts:21]: `Joi.AnySchema` is the widest schema type possible. Also, the `key` is totally useless because the same config would pass with them or not, unless you have different value schema for `desktop` / `mobile` than the other cases, which you don't

File: packages/docusaurus-theme-classic/src/theme/hooks/usePlatformValue.ts (hunk 5 above)
Comment (cerkiewny) [reply packages/docusaurus-theme-classic/src/theme/hooks/usePlatformValue.ts:18]: how should I handle ssr?

**Comment by Josh-Cena** [comment]
Type guards are not the simplest to design, and I'm fine to write it for you. I think a problem that prevents this from being as intuitive as it could is having too much in mind: we just want platform-dependent front matter for all string/boolean/number front matter, and even if we are to provide a `logo` field, it's going to be a string. We don't plan to port platform-dependency to docusaurus.config.js now, just in the front matter. With that said, I think it's safe to make assumptions like `platformDependent<T extends string | number | boolean>`

**Comment by Josh-Cena** [comment]
The `default` option doesn't make sense: the user only sees desktop or mobile with a clear-defined boundary (`996px` width). There's nothing as default.

**Comment by cerkiewny** [comment]
> 
> 
> The `default` option doesn't make sense: the user only sees desktop or mobile with a clear-defined boundary (`996px` width). There's nothing as default.

My understanding was that themes different than classic could handle more options so I wrote it with that in mind, but this indeed doesn't make sense if it is defined inside the theme-classic.

**Comment by Josh-Cena** [comment]
> My understanding was that themes different than classic could handle more options so I wrote it with that in mind, but this indeed doesn't make sense if it is defined inside the theme-classic.

The idea is that the plugin is quite lenient in terms of validation, and different themes would pick up different values that it recognizes. A user bears the burden herself if she decides to migrate to another theme with different screen size definitions. No need for a theme itself to worry about that.

File: packages/docusaurus-theme-classic/src/theme/hooks/usePlatformValue.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+function isSimpleValue(
+  option: string | number | boolean | Record<string, string | number | boolean>,
+): option is string | number | boolean {
+  return typeof option !== 'object';
+}
+
+function usePlatformValue<T extends string | number | boolean>(
+  option: PlatformDependentConfig<T> | undefined,
+): T | undefined {
+  const windowSize = useWindowSize();
```
Comment (slorber) [inline packages/docusaurus-theme-classic/src/theme/hooks/usePlatformValue.ts:21]: This is not possible to do this without introducing hydration layout shifts.

What is rendered on the server must stay the same once React hydrates on the client.

A UI element with `mobile: true` & `desktop: false` should not lead to the title magically appearing/disappearing after JS loads. 

You can't rely on the `windowSize` dynamic value, only on the config/frontmatter. This wouldn't work on the server. Remember also that Docusaurus may want to work better without React hydration in the future, so don't assume any JS will run on the client-side: it should work with JS disabled.

To avoid hydration layout shifts, we must play with CSS mediaqueries and also likely render twice the same element in the DOM with different mediaqueries.

This is not a simple problem to solve on a static website, but there are some solutions in the SSR land (https://artsy.github.io/blog/2019/05/24/server-rendering-responsively/)

For these reasons I'm not really in favor of supporting this mobile/desktop config for all the possible frontmatters, but only those for which there are a strong use-case.

**Review by slorber — CHANGES_REQUESTED** [review]

File: packages/docusaurus-theme-classic/src/theme/hooks/usePlatformValue.ts (hunk 7 above)
Comment (Josh-Cena) [reply packages/docusaurus-theme-classic/src/theme/hooks/usePlatformValue.ts:21]: There's a pretty strong use-case for most `hide_*` front matter. You mentioned the TOC one somewhere, and we agreed on the `hide_sidebar`. I think SSR can be handled gracefully here. Maybe the idea of `usePlatformValue` is still too generic

Comment (slorber) [reply packages/docusaurus-theme-classic/src/theme/hooks/usePlatformValue.ts:21]: We can definitively implement this but this `usePlatformValue` is a trap and using it will create layout shifts, we it must be done in another way, without this abstraction, and consider SSR properly everywhere such config is used.

> and we agreed on the hide_sidebar

Yes, and the sidebar is already rendered differently for mobile/desktop, the complexity is already there so adding this feature in such a case is less complex than introducing it for elements that render the same on mobile/desktop

Comment (cerkiewny) [reply packages/docusaurus-theme-classic/src/theme/hooks/usePlatformValue.ts:21]: I think the Slorber criticism is mostly of a fact that platform should not change during the visit on the site, for instance if you switch to "desktop mode" on mobile, the entire site should be refetched. And as such it doesn't make sense for the usePlatformValue to be a hook. But than we are going back to my original question, how should we determinate the platform.

Btw @slorber I think the media query are also based of off window size so you might experience layout shifts with them as well. Is "hydration layout shift" some specific problem?

Comment (cerkiewny) [reply packages/docusaurus-theme-classic/src/theme/hooks/usePlatformValue.ts:21]: There's no way to know which platform it is before arriving at the browser - you can actually determinate which platform it is before arriving at the browser, it is just not window size based. With high DPI phones the size based approach is a bit flawed anyway IMHO. Most modern browsers will send the platform indicating string to the server when fetching the site.

Comment (slorber) [reply packages/docusaurus-theme-classic/src/theme/hooks/usePlatformValue.ts:21]: > There's no way to know which platform it is before arriving at the browser

True, but the critical CSS is inlined into the static HTML pages, so the browser has all the CSS before painting your HTML page, and mediaqueries allow to hide/show elements even before the React code has finished loading, as soon as the first pixels are displayed to the user screen.

> The blog link Sebastien always sends: joshwcomeau.com/react/the-perils-of-rehydration

This post is about a different problem due to server/browser JSX mismatches, not really related to layout shifts.

---

TLDR: the feature you implement should rather work in a production build with JS disabled (there are chrome plugins to disable JS, and it is time-consuming to test as you really need to build the prod site).

A good source of inspiration is the `@theme/ThemedImage` component: you can see the complexity introduced to handle layout shifts:

```jsx
import React from 'react';
import clsx from 'clsx';

import useIsBrowser from '@docusaurus/useIsBrowser';
import useThemeContext from '@theme/hooks/useThemeContext';
import type {Props} from '@theme/ThemedImage';

import styles from './styles.module.css';

const ThemedImage = (props: Props): JSX.Element => {
  const isBrowser = useIsBrowser();
  const {isDarkTheme} = useThemeContext();
  const {sources, className, alt = '', ...propsRest} = props;

  type SourceName = keyof Props['sources'];

  const clientThemes: SourceName[] = isDarkTheme ? ['dark'] : ['light'];

  const renderedSourceNames: SourceName[] = isBrowser
    ? clientThemes
    : // We need to render both images on the server to avoid flash
      // See https://github.com/facebook/docusaurus/pull/3730
      ['light', 'dark'];

  return (
    <>
      {renderedSourceNames.map((sourceName) => (
        <img
          key={sourceName}
          src={sources[sourceName]}
          alt={alt}
          className={clsx(
            styles.themedImage,
            styles[`themedImage--${sourceName}`],
            className,
          )}
          {...propsRest}
        />
      ))}
    </>
  );
};

export default ThemedImage;
```

```
.themedImage {
  display: none;
}

html[data-theme='light'] .themedImage--light {
  display: initial;
}

html[data-theme='dark'] .themedImage--dark {
  display: initial;
}
```

You can see that we need to render in both light/dark version on the server, so that it does not show first the light version and then updates to the dark version.

(the `html[data-theme='light']` acts a bit like a mediaquery here because we set that data-attribute as soon as possible with js inlined in the static HTML files to avoid theme flashes => a problem that is quite similar to layout shifts in practice)

**Comment by slorber** [comment]
see https://github.com/facebook/docusaurus/issues/5805#issuecomment-1733350229
